decentralized-message-queue/src/DMQ/Protocol/LocalMsgNotification/Type.hs

```diff
@@ -0,0 +1,181 @@
+{-# LANGUAGE DataKinds                #-}
+{-# LANGUAGE DeriveFoldable           #-}
+{-# LANGUAGE FlexibleInstances        #-}
+{-# LANGUAGE PolyKinds                #-}
+{-# LANGUAGE StandaloneDeriving       #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeData                 #-}
+{-# LANGUAGE TypeFamilies             #-}
+
+-- | todo haddock
+module DMQ.Protocol.LocalMsgNotification.Type where
+
+import Control.DeepSeq
+import Data.Kind
+import Data.List.NonEmpty
+import Network.TypedProtocol.Core
+import Ouroboros.Network.Util.ShowProxy
+
+-- | There are some constraints of the protocol that are not captured in the
+-- types of the messages, but are documented with the messages. Violation
+-- of these constraints is also a protocol error. The constraints are intended
+-- to ensure that implementations are able to work in bounded space.
+--
+instance Protocol (LocalMsgNotification msg) where
+  -- TODO haddock
+  -- | The messages in the protocol.
+  --
+  -- There are two ways to ask for messages, blocking and
+  -- non-blocking (todo: blocking only should suffice?).
+  -- They otherwise have the same semantics.
+  --
+  data Message (LocalMsgNotification msg) from to where
+
+    -- | Request a list of messages from the server,
+    --
+    --
+    -- * The blocking case is used when the server has announced it has no
+    --   further messages to provide, and otherwise the non-blocking request
+    --   must be used.
+    -- TODO: this request should have the max number of messages that shall
+    -- be returned?
```

Comment:
```suggestion
    -- | Request a list of messages from the server,
    --
    -- * The blocking case is used when the server has announced it has no
    --   further messages to provide, and otherwise the non-blocking request
    --   must be used.
    --
    -- TODO: this request should have the max number of messages that shall
    -- be returned?
```